### PR TITLE
Parse .debug_ranges

### LINKED ISSUE #78: addr2line: use DW_AT_ranges
Related to #72 and #77.

We should use the `DW_AT_ranges` attribute to determine which addresses are contained in a compilation unit.

### PR DESCRIPTION
**Pull request by philipc**
Fixes #78

### CODE REVIEW

File: examples/addr2line.rs [hunk 1]
```diff
@@ -98,48 +102,97 @@ fn find_address<Endian>(debug_line: gimli::DebugLine<Endian>, units: &[Unit], ad
...
+        };
+
+        let high_pc = match entry.attr_value(gimli::DW_AT_high_pc) {
+            Some(gimli::AttributeValue::Addr(addr)) => addr,
+            Some(gimli::AttributeValue::Udata(size)) => low_pc.wrapping_add(size),
+            None => low_pc.wrapping_add(1),
+            _ => return None,
+        };
+
+        assert!(low_pc < high_pc);
```
Comment (fitzgen) [inline examples/addr2line.rs]: `debug_assert!` ?

File: examples/addr2line.rs [hunk 2]
```diff
@@ -98,48 +102,97 @@ fn find_address<Endian>(debug_line: gimli::DebugLine<Endian>, units: &[Unit], ad
...
+        Some(gimli::Range {
+            begin: low_pc,
+            end: high_pc,
+        })
+    }
+
     fn contains_address(&self, address: u64) -> bool {
-        self.high_pc == 0 || address >= self.low_pc && address <= self.high_pc
+        self.ranges.iter().find(|range| address >= range.begin && address < range.end).is_some()
```
Comment (fitzgen) [inline examples/addr2line.rs]: Nit: `self.ranges.iter().any(|range| address >= range.begin && address < range.end)`

File: src/constants.rs [hunk 3]
```diff
@@ -658,6 +658,6 @@ dw!(DwOp(u8) {
     DW_OP_implicit_value = 0x9e,
     DW_OP_stack_value = 0x9f,
 
-    // GNU extensions that are supported by our expression evaluator.
+// GNU extensions that are supported by our expression evaluator.
```
Comment (fitzgen) [inline src/constants.rs]: Unintentional?

File: src/lib.rs [hunk 4]
```diff
@@ -80,6 +80,8 @@
 //!   * [`DebugPubTypes`](./type.DebugPubTypes.html): The `.debug_pubtypes`
 //!   section.
 //!
+//!   * [`DebugRanges`](./struct.DebugRanges.html): The `.debug_ranges` section.
+//!
```
Comment (fitzgen) [inline src/lib.rs]: Thanks :)

File: src/ranges.rs [hunk 5]
```diff
@@ -0,0 +1,379 @@
...
+            0x00, 0x06, 0x01, 0x00, 0x00, 0x06, 0x01, 0x00,
+            0x00, 0x08, 0x01, 0x00, 0x00, 0x09, 0x01, 0x00,
+            // A range that starts at 0.
+            0x00, 0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00,
+            // A range that ends at -1.
+            0xff, 0xff, 0xff, 0xff, 0x00, 0x00, 0x00, 0x00,
+            0x00, 0x00, 0x00, 0x00, 0xff, 0xff, 0xff, 0xff,
+            // A range end.
+            0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+            0xff,
```
Comment (fitzgen) [inline src/ranges.rs]: Don't want to use `test-assembler` for constructing these?

**Comment by fitzgen** [comment]
LGTM, but I didn't read up on the exact form of `.debug_ranges` in the standard.

File: examples/addr2line.rs (hunk 1 above)
Comment (philipc) [inline examples/addr2line.rs]: No, this depends on the input file. Added a comment.

File: examples/addr2line.rs (hunk 2 above)
Comment (philipc) [inline examples/addr2line.rs]: Done.

File: src/constants.rs (hunk 3 above)
Comment (philipc) [inline src/constants.rs]: Yes, this is due to running rustfmt. Will fix that in a separate commit.

File: src/ranges.rs (hunk 5 above)
Comment (philipc) [inline src/ranges.rs]: Habit :) Done.

**Comment by philipc** [comment]
Thanks for the review.

I also noticed that coveralls isn't showing the new file.

File: src/ranges.rs [hunk 6]
```diff
@@ -0,0 +1,453 @@
+use endianity::{Endianity, EndianBuf};
```
Comment (tromey) [inline src/ranges.rs]: Other files have `#![deny(missing_docs)]` at the top, which seems like a good idea.

Comment (philipc) [inline src/ranges.rs]: It's a good idea, but I think this should be done as a crate wide setting, and it seems that the `#![deny(missing_docs)]` in `lib.rs` is enough.

**Pull request merged**